### ISSUE TITLE
Aligning interfaces

### ISSUE THREAD

**Issue by LoyVanBeek** [RoboCupAtHome/RuleBook#482]
On behalf of @Reinzor, I also think the league would progress faster with more software sharing and being able to exchange and re-use components. 
Having been TC, I can't enforce this nor should it be enforced, but encouraging teams to collaborate more should be a goal. 

> - trying to work on a high level interface with defined components with interfaces
> - this could be a joint effort between the more experienced teams
> - everybody should list their components in the tdp and make these available via git (Also see https://github.com/RoboCupAtHome/RuleBook/issues/479 and especially https://github.com/RoboCupAtHome/AtHomeCommunityWiki/wiki/Software#Team_software_overview)
> - people should log the calls over these interfaces (most probably topics)
> - this can be used by tools created by the TC for scoring: so fully automated objective scoring instead of the current subjective scoring system
> - there should be some challenges where people can show there new developments freely but this should be separately scored, like a technical challenge 
> - this architecture should be updated once x years 
> 
> I thinks this could really help the league developing, share components and create a startup software suite for new teams in order to get them started quickly.

**Comment by maximest-pierre** [comment]
The way I see it would be something that is ROS agnostic that we could in theory allow other leagues and robots to join us in challenges. It could be something simple like an http api with a set of commands and protocol.

**Comment by LoyVanBeek** [comment]
Ideally (for me) you wrap ROS-independent libraries in proper ROS and maybe other wrappers. The wrapper makes I can just git clone, maybe compile and rosrun something. If I want to use something else, I have to do some (de)serialization and I don't like to write that for every library. 
ROS is a de-facto standard and while it isn't perfect, it's very handy. More handy than having lots of incompatible libraries that use slightly different, incompatible datatypes.

**Comment by kyordhel** [comment]
I think we should set guidelines for that. To me, an ideal solution is as follows.

The solution has three components
1. A static/shared library 
  - The library can be compiled and tested standalone
  - All interfacing components/functions/wrappers are properly documented
  - The url of the library is provided as part of the documentation (e.g. github repo)
  - Preferably, the library is open source
2. An standalone test program
  - Within the program, the core interface code are used and its use explained
  - Tunning parameters and other operative settings are exposed and detailed here
  - The program details the library prerequisites and allows an out-of-the-box test-and-run
3. An example ROS package
  - Clone, install and run.
  - Your packages just need to refer to the topic services of the interface

So basically I can test your contribution in less than 5 minutes and integrate in my environment in no time. That I can make it actually work smooth will require further tunning of the system, but once it's running one starts adjusting parameters.

Further on, we can work in some sort of standard or convention to name topics/services so components in @home become interchangeable.

**Comment by maximest-pierre** [comment]
I think that that package could be a good test case. There is a lot of different wrapper for OpenPose for ROS on github (Most are crappy and Proof of Concept) but none of them support 3D skeletons as far as I can tell. This will also level the playing field for new teams and make the competition more focus on innovation instead of just integrating existing software. 

CC. @Philippe117 @JeffCousineau

**Comment by reinzor** [comment]
I do agree with all the points mentioned above. However, I think we should just simply stick with a ROS API in order to not complex things. People are using ROS because they can try "standard components" easily. If we are going to design complex interfaces on a python or cpp level (or others), I think we are not focusing on what we are trying to achieve here: **a more standardized @home robot architecture that enables interchangeability of modules**. As @LoyVanBeek mentioned, a lot of teams have the same modules in their system ([illustrated here](https://github.com/RoboCupAtHome/AtHomeCommunityWiki/wiki/Software#Team_software_overview) - otherwise people could not fill in a table with "standard" columns). Let's try to define interfaces for these columns so we can actually use each others stuff :).

**Comment by kyordhel** [comment]
Close but not cigar. Whatever code is being shared mot likely has some valuable research behind. This research is prone (and worthy) of being re-addressed by other groups, making @Home a perfect place not only sharing engineering solutions but also fostering league-wide scientific collaboration. That's the main reason I am requesting a standalone library (or set of libraries), so teams can test the library a a research component outside ROS. Further, I wouldn't like to give any additional advantage to ROS users over non-ROS users whose research is also valuable

**Comment by airglow** [comment]
I support @reinzor. ROS is de facto standard, and can also be integrated by other systems as well and the other way around. We should keep it simple.  When we keep to standard messages that are used by ROS already, the usage on other robots is mostly straight forward (we did that with our stuff on conferences, competitions that we attended without a robot or where our robot broke). We can learn a lot from the machine learning community, where simple instructions and pre-trained models make the use and integration often straight forward.
We will get back to that once the new project is started. I plan to publish all packages and make them installable for kinetic directly.

**Comment by kyordhel** [comment]
@LoyVanBeek Bump!

**Comment by johaq** [comment]
We have a lot of (very old) topics on standardization. Nothing every really happens on this front because it is difficult to enforce this in the rules and teams are mostly working for themselves at the competition because they want to get their stuff working. Maybe we could host a workshop as part of the symposium to discuss this further? That is pretty much the only point in time where we have a lot of teams in the same place and they are not in competition mode.
